File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Unit/Monster/ViewModel/TheBottom/Sentry_ViewModel.ts
import { Spine } from "csharp";
import { Mathf } from "../../../../../../../Core/Module/Math/Mathf";
import VFXAction from "../../../../Action/Animations/VFXAction";
import { ApplyPowerAction } from "../../../../Action/Common/ApplyPowerAction";
import DamageAction from "../../../../Action/Common/DamageAction";
import { DoFuncAction } from "../../../../Action/Common/DoFuncAction";
import { MakeTempCardInDiscardAction } from "../../../../Action/Common/MakeTempCardInDiscardAction";
import { RollMoveAction } from "../../../../Action/Common/RollMoveAction";
import { SFXAction } from "../../../../Action/Utility/SFXAction";
import { Dazed } from "../../../../Cards/Status/Dazed";
import { AttackEffect } from "../../../../DataDefine/AttackEffect";
import Color from "../../../../DataDefine/Color";
import DamageInfo from "../../../../DataDefine/DamageInfo";
import { DamageType } from "../../../../DataDefine/DamageType";
import DungeonManager from "../../../../DungeonManager";
import { BorderFlashEffect } from "../../../../Effect/Combat/BorderFlashEffect";
import { ShockWaveEffect, ShockWaveType } from "../../../../Effect/Combat/ShockWaveEffect";
import { SmallLaserEffect } from "../../../../Effect/Combat/SmallLaserEffect";
import { OnPreBattleMessage } from "../../../../Events/OnPreBattleMessage";
import { ArtifactPower } from "../../../../Power/ArtifactPower";
import { AbstractMonster } from "../../AbstractMonster";
import { Intent } from "../../Intent";

export class GremlinWizard_ViewMode extends AbstractMonster{
    private _dazedAmt = 0;
    Initialize() {
        if(DungeonManager.Inst.AdvanceLevel >= 18){
            this._dazedAmt = 3;
        }
        if(DungeonManager.Inst.AdvanceLevel >= 3) {
            this.DamageInfo.push(new DamageInfo(this,10));
        }else{
            this._dazedAmt = 2;
            this.DamageInfo.push(new DamageInfo(this,9));
        }
        let trackEntry = <Spine.TrackEntry>this.SetAnimation("idle",true);
        trackEntry.TrackTime = trackEntry.TrackEnd * Mathf.Random();
        trackEntry.TimeScale = 2;
        let animation = <Spine.Unity.SkeletonAnimation>this.DisplayObject.Self;
        let data = animation.SkeletonDataAsset.GetAnimationStateData();
        data.SetMix("idle","attack",0.1);
        data.SetMix("idle","spaz1",0.1);
        data.SetMix("idle","hit",0.1);
    }

    PreBattleAction(msg: OnPreBattleMessage) {
        super.PreBattleAction(msg);
        this.AddToBot(new ApplyPowerAction(this,this,new ArtifactPower(),1));
    }

    GetMove(num: number) {
        let room = DungeonManager.Inst.CurrentRoom;
        if(room.Round == 1){
            if(room.Monsters.IndexOf(this) % 2){
                this.SetMove(3,Intent.DEBUFF);
            }
            else{
                this.SetMove(4,Intent.ATTACK,{damage:this.DamageInfo[0]});
            }
        }
        if(this.LastMove(4)){
            this.SetMove(3,Intent.DEBUFF);
        }else{
            this.SetMove(4,Intent.ATTACK,{damage:this.DamageInfo[0]});
        }
    }

    TakeTurn() {
        switch(this.MoveIndex){
            case 3:{
                this.AddToBot(new SFXAction("THUNDERCLAP"));
                let myHitBox = this.GetHitBox();
                this.AddToBot(new VFXAction(new ShockWaveEffect(myHitBox.CX,myHitBox.CY,Color.Royal,ShockWaveType.ADDITIVE),0.5));
                this.AddToBot(new DoFuncAction(()=>this.PlayFastShake(0.6),0.2));
                this.AddToBot(new MakeTempCardInDiscardAction(new Dazed(),this._dazedAmt));
                break;
            }
            case 4:{
                this.AddToBot(new DoFuncAction(this.Attack));
                this.AddToBot(new SFXAction("ATTACK_MAGIC_BEAM_SHORT",0.5));
                this.AddToBot(new VFXAction(new BorderFlashEffect(Color.Sky)))
                let playerHitBox = DungeonManager.Inst.Player.GetHitBox();
                let myHitBox = this.GetHitBox();
                this.AddToBot(new VFXAction(new SmallLaserEffect(playerHitBox.CX,playerHitBox.CY,myHitBox.CX,myHitBox.CY),0.3));
                this.AddToBot(new DamageAction(DungeonManager.Inst.Player,this.DamageInfo[0],AttackEffect.NONE));
                break;
            }
        }
        this.AddToBot(new RollMoveAction(this));
    }

    private Attack(){
        this.SetAnimation("attack");
        this.AddAnimation("idle",true);
    }
    
    Damage(info: DamageInfo) {
        super.Damage(info);
        if(info.TotalDamage > 0 && info.DamageType != DamageType.THORNS){
            this.SetAnimation("hit");
            this.AddAnimation("idle",true);
        }
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Unit/Monster/ViewModel/TheBottom/Sentry_ViewModel.cs
using Core.Utils;
using ModTheSpire.Card.Status;
using ModTheSpire.DataDefine;
using ModTheSpire.Dungeon;
using ModTheSpire.Effect;
using ModTheSpire.Effect.Combat;
using ModTheSpire.Effect.Other;
using ModTheSpire.Event.Ids;
using ModTheSpire.Power.Buff;
using ModTheSpire.Power.Debuff;
using ModTheSpire.Unit.Monster.Helper;
using Spine;
using Spine.Unity;
using TCG.Config;
using UnityEngine;

namespace ModTheSpire.Unit.Monster.ViewModel.TheBottom
{
    public class Sentry_ViewModel : AbstractMonster
    {
        private int mDazedAmt = 0;

        protected override void Initialize()
        {
            if (DungeonManager.Inst.AdvanceLevel >= 18)
            {
                mDazedAmt = 3;
            }

            if(DungeonManager.Inst.AdvanceLevel >= 3) {
                this.DamageInfo.Add(new DamageInfo(this,10));
            }else{
                mDazedAmt = 2;
                this.DamageInfo.Add(new DamageInfo(this,9));
            }
            var trackEntry = (TrackEntry)this.DisplayObject.SetAnimation("idle", true);
            trackEntry.TrackTime = trackEntry.TrackEnd * Random.value;
            trackEntry.TimeScale = 2;
            var animation = (SkeletonAnimation)this.DisplayObject.Self;
            var data = animation.SkeletonDataAsset.GetAnimationStateData();
            data.SetMix("idle","attack",0.1f);
            data.SetMix("idle","spaz1",0.1f);
            data.SetMix("idle","hit",0.1f);
        }

        protected override void PreBattleAction(OnPreBattleMessage msg)
        {
            base.PreBattleAction(msg);
            this.AddToBot(new ApplyPowerAction(this,this,new ArtifactPower(),1));

        }

        protected override void GetMove(int num)
        {
            var room = DungeonManager.Inst.CurrentRoom;
            if(room.Round == 1){
                if(room.Monsters.IndexOf(this) % 2 == 0){
                    this.SetMove(3,Intent.DEBUFF);
                }
                else{
                    this.SetMove(4,Intent.ATTACK,this.DamageInfo[0]);
                }
            }
            if(this.LastMove(4)){
                this.SetMove(3,Intent.DEBUFF);
            }else{
                this.SetMove(4,Intent.ATTACK,this.DamageInfo[0]);
            }
        }

        protected override void TakeTurn()
        {
            switch(this.MoveIndex){
                case 3:{
                    this.AddToBot(new SFXAction(SoundMaster.THUNDERCLAP));
                    var myHitBox = this.DisplayObject.GetHitBox();
                    this.AddToBot(new VFXAction(
                        new ShockWaveEffect(myHitBox.center.x, myHitBox.center.y, MTSColor.Royal.Clone(), ShockWaveType.ADDITIVE), 500));
                    this.AddToBot(new DoFuncAction(()=>this.DisplayObject.PlayFastShake(0.6f),200));
                    this.AddToBot(new MakeTempCardInDiscardAction(new Dazed(),this.mDazedAmt));
                    break;
                }
                case 4:{
                    this.AddToBot(new DoFuncAction(this.Attack));
                    this.AddToBot(new SFXAction(SoundMaster.ATTACK_MAGIC_BEAM_SHORT,500));
                    this.AddToBot(new VFXAction(new BorderFlashEffect(MTSColor.Sky.Clone())));
                    var playerHitBox = DungeonManager.Inst.Player.DisplayObject.GetHitBox();
                    var myHitBox = this.DisplayObject.GetHitBox();
                    this.AddToBot(new VFXAction(
                        new SmallLaserEffect(playerHitBox.center.x, playerHitBox.center.y, myHitBox.center.x, myHitBox.center.y), 300));
                    this.AddToBot(new DamageAction(DungeonManager.Inst.Player,this.DamageInfo[0]));
                    break;
                }
            }
            this.AddToBot(new RollMoveAction(this));
        }

        
        private void Attack(){
            this.DisplayObject.SetAnimation("attack");
            this.DisplayObject.SetAnimation("idle",true);
        }

        public override void Damage(DamageInfo info)
        {
            base.Damage(info);
            if(info.Total > 0 && info.Type != DamageType.THORNS){
                this.DisplayObject.SetAnimation("hit");
                this.DisplayObject.AddAnimation("idle",true);
            }
        }

        public Sentry_ViewModel(MonsterInfo info) : base(info)
        {
        }
    }
}